### PR TITLE
Traefik route endpoint

### PR DESCRIPTION
**Pull request by PietroPasotti**
## Feature
This PR adds an integration with the Traefik Route charm (cf. https://github.com/PietroPasotti/traefik-route-k8s-operator), enabling users to configure Traefik on a per-relation basis, overcoming the limitations of ingress-per-unit and ingress-per-app, making Traefik a lot more general.

The role of Traefik in this integration is very limited: it simply has to accept a yaml config file and write it to the folder where traefik will load it up. Therefore the changes to the charm codebase are rather limited. 

## Testing Instructions
Aside from running the utests, you can test this in combination with TraefikRoute and prometheus-k8s in a live deployment (itests pending).

```
juju deploy traefik-route-k8s --channel beta
```

## Release Notes
* Added `traefik_route_k8s/v0/traefik_route.py` lib
* Added `traefik_route` integration in the Traefik charm.

### CODE REVIEW

File: lib/charms/traefik_route_k8s/v0/traefik_route.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+cd some-charm
+charmcraft fetch-lib charms.traefik_route_k8s.v0.traefik_route
+```
+
+```yaml
+requires:
+    traefik_route:
+        interface: traefik_route
+        limit: 1
+```
```
Comment (jnsgruk) [inline lib/charms/traefik_route_k8s/v0/traefik_route.py:24]: As I understand it - the design was such that the workload needing ingress would only ever need to communicate on the `ingress` or `ingress_per_unit` relation. 

This interface should only be present between traefik-k8s and traefik-route-k8s, right?

File: lib/charms/traefik_route_k8s/v0/traefik_route.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    # ...
+    self.ingress_per_unit = TraefikRouteProvider(self)
+    self.traefik_route = TraefikRouteRequirer(self)
+
+    self.framework.observe(
+        self.ingress_per_unit.on.request, self.traefik_route.relay
+    )
+    self.framework.observe(
+        self.traefik_route.on.response, self.ingress_per_unit.respond
+    )
```
Comment (jnsgruk) [inline lib/charms/traefik_route_k8s/v0/traefik_route.py:43]: As above - I think something was lost in translation here.

File: lib/charms/traefik_route_k8s/v0/traefik_route.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        if self.is_ready(event.relation):
+            # todo check data is valid here?
+            self.on.ready.emit(event.relation)
+
+    @staticmethod
+    def is_ready(relation: Relation) -> bool:
+        """Whether TraefikRoute is ready on this relation: i.e. the remote app shared the config."""
+        return 'config' in relation.data[relation.app]
+
+    @staticmethod
```
Comment (jnsgruk) [inline lib/charms/traefik_route_k8s/v0/traefik_route.py:127]: Do these need to be static?

File: lib/charms/traefik_route_k8s/v0/traefik_route.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+class TraefikRouteRequirer(Object):
+    """Wrapper for the requirer side of traefik-route.
+
+    traefik_route will publish to the application databag an object like:
+    {
+        'config': <Traefik_config>
+    }
+
+    'ingress' is provided by the ingress end-user via ingress_per_unit,
```
Comment (jnsgruk) [inline lib/charms/traefik_route_k8s/v0/traefik_route.py:142]: `ingress` is not included in your example above - should it be?

File: src/charm.py [hunk 5]
```diff
@@ -277,82 +288,86 @@ def _process_ingress_relation(self, relation: Relation):
...
-                    if self.unit.is_leader():
-                        request.respond(unit, unit_url)
+            self._provide_ingress_per_unit(relation)
 
-                if unit_config:
-                    always_merger.merge(config, unit_config)
+    def _provide_routed_ingress(self, relation: Relation):
+        """Provide ingress to a unit related through TraefikRoute."""
+        if not self.traefik_route.is_ready(relation):
+            logger.info(f"TR not ready on {relation}")
```
Comment (jnsgruk) [inline src/charm.py:307]:
```suggestion
            logger.info("%s not ready on %s", self.app.name, relation)
```

**Review by jnsgruk — CHANGES_REQUESTED** [review]

File: lib/charms/traefik_route_k8s/v0/traefik_route.py (hunk 3 above)
Comment (PietroPasotti) [reply lib/charms/traefik_route_k8s/v0/traefik_route.py:127]: No, but my pycharm linter complains if they aren't. Do they need to not be static?

File: lib/charms/traefik_route_k8s/v0/traefik_route.py (hunk 2 above)
Comment (PietroPasotti) [reply lib/charms/traefik_route_k8s/v0/traefik_route.py:43]: Uh yes, copy-pasted from the IPU library a bit too freely. On it

File: lib/charms/traefik_route_k8s/v0/traefik_route.py (hunk 3 above)
Comment (jnsgruk) [reply lib/charms/traefik_route_k8s/v0/traefik_route.py:127]: Just seems odd to me, but I guess it's not a huge problem.

**Review by jnsgruk — APPROVED** [review]

**Pull request closed without merging**